magefiles/run.go

```diff
@@ -0,0 +1,209 @@
+//go:build mage
+// +build mage
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+)
+
+// Create and run Kubernetes cluster with two nodes, Linux and Windows.
+func Run() error {
+	mg.SerialDeps(startup, checkVagrant, checkClusterNotRunning)
+
+	var err error
+
+	log.Println("Creating .lock directory")
+	_, err = os.Stat(".lock")
+	if !os.IsNotExist(err) {
+		err = os.RemoveAll(".lock")
+		if err != nil {
+			return err
+		}
+	}
+	err = os.Mkdir(".lock", os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// Linux node //////////////////////////////////////////////////////////////
+	log.Println("Creating control plane Linux node")
+	err = sh.Run("vagrant", "up", "controlplane")
+	if err != nil {
+		return err
+	}
+
+	mustSetVagrantPrivateKeyPermissions()
+
+	err = sh.Run("vagrant", "status")
+	if err != nil {
+		return err
+	}
+
+	// Windows node ////////////////////////////////////////////////////////////
+	log.Println("Creating worker Windows node")
+	log.Println("##########################################################")
+	log.Println("Retry vagrant up if the first time the Windows node failed")
+	log.Println("##########################################################")
+
+	const maxAttempts = 10 // TODO: How many to allow? Make configurable via variables.yaml?
+	for i := 1; i < maxAttempts; i++ {
+		log.Printf("vagrant status winw1 - attempt %d of %d", i, maxAttempts)
+		output, err := sh.Output("vagrant", "status", "winw1")
+		if err != nil {
+			return err
+		}
+		status := extractMachineStatus(output, "winw1")
+		log.Println("winw1", status)
+		if strings.Contains(status, "running") {
+			break
+		}
+		log.Printf("vagrant up winw1 - attempt %d of %d", i, maxAttempts)
+		err = sh.Run("vagrant", "up", "winw1")
+		if err != nil {
+			return err
+		}
+	}
+
+	mustSetVagrantPrivateKeyPermissions()
+
+	for i := 1; i < maxAttempts; i++ {
+		log.Printf("kubectl get nodes | grep winw1  - attempt %d of %d", i, maxAttempts)
+		output, err := sh.Output("vagrant", "ssh", "controlplane", "-c", "kubectl get nodes")
+		if err != nil {
+			return err
+		}
+		log.Println(output)
+		if strings.Contains(output, "winw1") {
+			break
+		}
+		log.Printf("vagrant provision winw1 - attempt %d of %d", i, maxAttempts)
+		err = sh.Run("vagrant", "provision", "winw1")
+		if err != nil {
+			return err
+		}
```

Comment:
same here.

Reply:
https://github.com/kubernetes-sigs/sig-windows-dev-tools/pull/264#discussion_r1181300214

There is multiple testing threads going on and the whole vagrant sequence for the Windows node is going to be rewritten and simplified as those loops seem to be a workaround for issues with flaky Vagrant box. Once we have improved Vagrant box with Windows, which likely won't need it the loops, then we can try the sleep tweak.